specification.md

```diff
@@ -0,0 +1,1590 @@
+
+Since 0.4.0 changes some of the LSIF aspects more deeply an old 0.3.x version of the specification is available [here](https://github.com/microsoft/language-server-protocol/blob/master/indexFormat/versions/specification-0-3-x.md)
+
+## <a href="#lsifMotivation" name="lsifMotivation" class="anchor">Motivation</a>
+
+Principal design goals:
+
+- The format should not imply the use of a certain persistence technology.
+- The data defined should be modeled as closely as possible to the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) to make it possible to serve the data through the LSP without further transformation.
+- The data stored is result data usually returned from a LSP request. The dump doesn't contain any program symbol information nor does the LSIF define any symbol semantics (e.g. where a symbol is defined or referenced or when a method overrides another method). The LSIF therefore doesn't define a symbol database. Please note that this is consistent with the LSP itself which doesn't define any symbol semantics either.
+- The output format will be based on JSON as with the LSP.
+
+LSP requests that are good candidates to be supported in LSIF are:
+
+- [`textDocument/documentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_documentSymbol)
+- [`textDocument/foldingRange`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_foldingRange)
+- [`textDocument/documentLink`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_documentLink)
+- [`textDocument/definition`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_definition)
+- [`textDocument/declaration`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_declaration)
+- [`textDocument/typeDefinition`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#textDocument_typeDefinition)
+- [`textDocument/hover`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15#textDocument_hover)
+- [`textDocument/references`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15#textDocument_references)
+- [`textDocument/implementation`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15#textDocument_implementation)
+- [`workspace/symbol`](https://microsoft.github.io/language-server-protocol/specifications/specification-current#workspace_symbol)

```

Comment:
To be clear, this was introduced in https://github.com/sourcegraph/lsif-protocol/pull/4/commits/2b43f2f263a5c7dd2672f8eec6ddc4196f55336d and isn't currently part of the LSIF specification.

@sqs Have you filed any issue or pull request against the LSIF specification, or is this going to be an ad-hoc extension? We've yet to have the discussion as a team about breaking from the specification.

Reply:
No, this would be an ad-hoc extension. It was a prototype project and I didn't give thought to the consequences of that beyond making something reasonably not awful that would work.

Reply:
I think we'd want to upstream this to the specification eventually. In my experience, however, proposed changes to specs are more likely to get through if there is already a demonstrable use case.